File: ts/packages/over-redis/src/client-transport.ts
import { randomUUID } from 'node:crypto';
import type { Redis } from 'ioredis';
import {
  parseEnvelope, EnvelopeKind, ClamatorTransportError,
  type Transport, type SendOptions, type Dispatcher,
} from '@clamator/protocol';
import { commandStream, replyStream } from './keys.js';

export interface ClientTransportOptions {
  redis: Redis;
  keyPrefix: string;
  instanceId?: string;
  replyStreamMaxLen?: number;
  defaultTimeoutMs?: number;
}

interface Pending {
  resolve: (env: Record<string, unknown>) => void;
  reject: (err: unknown) => void;
  timer: ReturnType<typeof setTimeout>;
}

export class ClientRedisTransport implements Transport {
  readonly instanceId: string;
  private readonly replyStream: string;
  private state: 'idle' | 'started' | 'stopped' = 'idle';
  private pending = new Map<string, Pending>();
  private replyLoop: Promise<void> | null = null;
  private replyLoopAbort = false;
  // Dedicated connection for the blocking XREAD reply loop, so that
  // xadd calls in send() are not queued behind the blocking read.
  private replyRedis: Redis | null = null;

  constructor(private readonly opts: ClientTransportOptions) {
    this.instanceId = opts.instanceId ?? randomUUID();
    this.replyStream = replyStream(opts.keyPrefix, this.instanceId);
  }

  async registerService(_name: string, _dispatch: Dispatcher): Promise<void> {
    throw new Error('client transport cannot host services');
  }

  async send(env: Record<string, unknown>, sendOpts: SendOptions): Promise<Record<string, unknown>> {
    if (this.state !== 'started')
      throw new ClamatorTransportError(`transport not started (state=${this.state})`);
    const parsed = parseEnvelope(env);
    if (parsed.kind !== EnvelopeKind.Request)
      throw new ClamatorTransportError('send requires a request envelope');
    const idStr = String(parsed.id);
    return new Promise<Record<string, unknown>>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(idStr);
        reject(new ClamatorTransportError('call timeout'));
      }, sendOpts.timeoutMs);
      this.pending.set(idStr, { resolve, reject, timer });
      void this.opts.redis.xadd(
        commandStream(this.opts.keyPrefix, parsed.service),
        '*',
        'type', 'rpc',
        'envelope', JSON.stringify(env),
        'reply-to', this.replyStream,
      ).catch(err => {
        const p = this.pending.get(idStr);
        if (!p) return;
        this.pending.delete(idStr);
        clearTimeout(p.timer);
        p.reject(new ClamatorTransportError('xadd failed', err));
      });
    });
  }

  async notify(env: Record<string, unknown>): Promise<void> {
    if (this.state !== 'started')
      throw new ClamatorTransportError(`transport not started (state=${this.state})`);
    const parsed = parseEnvelope(env);
    if (parsed.kind !== EnvelopeKind.Notification)
      throw new ClamatorTransportError('notify requires a notification envelope');
    await this.opts.redis.xadd(
      commandStream(this.opts.keyPrefix, parsed.service),
      '*',
      'type', 'rpc',
      'envelope', JSON.stringify(env),
      // no reply-to
    );
  }

  async start(): Promise<void> {
    if (this.state === 'stopped') throw new ClamatorTransportError('transport has been stopped');
    if (this.state === 'started') return;
    this.state = 'started';
    this.replyLoopAbort = false;
    // Use a dedicated duplicate connection for the blocking XREAD so that
    // xadd commands in send() are not queued behind the blocking read.
    this.replyRedis = this.opts.redis.duplicate();
    this.replyLoop = this.runReplyLoop().catch(err => {
      // surface fatal loop errors
      console.error('[clamator/over-redis] reply loop fatal:', err);
    });
  }

  async stop(): Promise<void> {
    if (this.state !== 'started') { this.state = 'stopped'; return; }
    this.replyLoopAbort = true;
    // Disconnect the reply connection to unblock the XREAD immediately.
    try { this.replyRedis?.disconnect(); } catch { /* ignore */ }
    try { await this.replyLoop; } catch { /* ignore */ }
    for (const [id, p] of this.pending.entries()) {
      clearTimeout(p.timer);
      p.reject(new ClamatorTransportError('transport stopped'));
      this.pending.delete(id);
    }
    try { await this.opts.redis.del(this.replyStream); } catch { /* best effort */ }
    try { await this.replyRedis?.quit(); } catch { /* best effort */ }
    this.replyRedis = null;
    this.state = 'stopped';
  }

  private async runReplyLoop(): Promise<void> {
    // Use '0-0' rather than '$' so replies are never missed between iterations.
    // The reply stream is unique per client instance, so reading from the
    // beginning is always correct and safe.
    let lastId = '0-0';
    while (!this.replyLoopAbort) {
      try {
        const result = await this.replyRedis!.xread('BLOCK', 1000, 'STREAMS', this.replyStream, lastId);
        if (!result) continue;
        for (const [, entries] of result as [string, [string, string[]][]][]) {
          for (const [entryId, fields] of entries) {
            lastId = entryId;
            const idx = fields.indexOf('envelope');
            if (idx < 0) continue;
            const json = fields[idx + 1] ?? '';
            let parsed: Record<string, unknown>;
            try { parsed = JSON.parse(json) as Record<string, unknown>; }
            catch { continue; }
            const replyId = String(parsed.id);
            const pending = this.pending.get(replyId);
            if (!pending) continue;  // late or stranger reply
            this.pending.delete(replyId);
            clearTimeout(pending.timer);
            pending.resolve(parsed);
          }
        }
      } catch (err) {
        if (this.replyLoopAbort) return;
        await new Promise(r => setTimeout(r, 100));
      }
    }
  }
}

File: py/packages/over-redis/src/clamator_over_redis/client_transport.py
from __future__ import annotations
import asyncio
import json
import os
import uuid
from typing import Any

from redis.asyncio import Redis

from clamator_protocol import (
    parse_envelope, ClamatorTransportError, RequestEnvelope, NotificationEnvelope,
    Dispatcher,
)
from .keys import command_stream, reply_stream


class ClientRedisTransport:
    def __init__(
        self, *, redis: Redis | None = None, redis_url: str | None = None,
        key_prefix: str,
        instance_id: str | None = None,
        default_timeout_ms: int = 30_000,
    ) -> None:
        if redis is not None and redis_url is not None:
            raise ClamatorTransportError("provide either `redis` or `redis_url`, not both")
        if redis is not None:
            self._redis = redis
            self._owns_redis = False
        else:
            url = redis_url or os.environ.get("REDIS_URL") or "redis://localhost:6379"
            self._redis = Redis.from_url(url)
            self._owns_redis = True
        self._key_prefix = key_prefix
        self.instance_id = instance_id or str(uuid.uuid4())
        self._reply_stream = reply_stream(key_prefix, self.instance_id)
        self._default_timeout = default_timeout_ms / 1000
        self._state = "idle"
        self._pending: dict[str, asyncio.Future[dict[str, Any]]] = {}
        self._reply_loop_task: asyncio.Task[Any] | None = None

    async def register_service(self, name: str, dispatch: Dispatcher) -> None:
        raise NotImplementedError("client transport cannot host services")

    async def send(self, env: dict[str, Any], *, timeout: float) -> dict[str, Any]:
        if self._state != "started":
            raise ClamatorTransportError(f"transport not started (state={self._state})")
        parsed = parse_envelope(env)
        if not isinstance(parsed, RequestEnvelope):
            raise ClamatorTransportError("send requires a request envelope")
        loop = asyncio.get_running_loop()
        fut: asyncio.Future[dict[str, Any]] = loop.create_future()
        self._pending[str(parsed.id)] = fut
        try:
            await self._redis.xadd(
                command_stream(self._key_prefix, parsed.service),
                {"type": "rpc", "envelope": json.dumps(env), "reply-to": self._reply_stream},
            )
        except Exception as e:
            self._pending.pop(str(parsed.id), None)
            raise ClamatorTransportError("xadd failed", cause=e) from e
        try:
            return await asyncio.wait_for(fut, timeout=timeout)
        except asyncio.TimeoutError as e:
            self._pending.pop(str(parsed.id), None)
            raise ClamatorTransportError("call timeout") from e

    async def notify(self, env: dict[str, Any]) -> None:
        if self._state != "started":
            raise ClamatorTransportError(f"transport not started (state={self._state})")
        parsed = parse_envelope(env)
        if not isinstance(parsed, NotificationEnvelope):
            raise ClamatorTransportError("notify requires a notification envelope")
        await self._redis.xadd(
            command_stream(self._key_prefix, parsed.service),
            {"type": "rpc", "envelope": json.dumps(env)},
        )

    async def start(self) -> None:
        if self._state == "stopped":
            raise ClamatorTransportError("transport has been stopped")
        if self._state == "started":
            return
        self._state = "started"
        self._reply_loop_task = asyncio.create_task(self._reply_loop())

    async def stop(self) -> None:
        if self._state != "started":
            self._state = "stopped"
            return
        self._state = "stopped"
        if self._reply_loop_task:
            self._reply_loop_task.cancel()
            try:
                await self._reply_loop_task
            except asyncio.CancelledError:
                pass
        for fut in list(self._pending.values()):
            if not fut.done():
                fut.set_exception(ClamatorTransportError("transport stopped"))
        self._pending.clear()
        try:
            await self._redis.delete(self._reply_stream)
        except Exception:
            pass
        if self._owns_redis:
            try:
                await self._redis.aclose()
            except Exception:
                pass

    async def _reply_loop(self) -> None:
        last_id = "0"
        while True:
            try:
                results = await self._redis.xread({self._reply_stream: last_id}, block=1000)
                if not results:
                    continue
                for _stream, entries in results:
                    for entry_id, fields in entries:
                        last_id = entry_id
                        env_field = fields.get(b"envelope") or fields.get("envelope")
                        if env_field is None:
                            continue
                        if isinstance(env_field, bytes):
                            env_field = env_field.decode()
                        try:
                            parsed = json.loads(env_field)
                        except Exception:
                            continue
                        rid = str(parsed.get("id"))
                        fut = self._pending.pop(rid, None)
                        if fut and not fut.done():
                            fut.set_result(parsed)
            except asyncio.CancelledError:
                return
            except Exception:
                await asyncio.sleep(0.1)
